internal/store/builder.go

```diff
@@ -74,7 +75,8 @@ func NewBuilder() *Builder {
 
 // WithMetrics sets the metrics property of a Builder.
 func (b *Builder) WithMetrics(r prometheus.Registerer) {
```

Comment:
Hmm but if we don't use WithSharding this is still created, I would rather make it dependant on the Sharding enabled here.

Reply:
Fair point! I should have mentioned that in the original description. My understanding is that KSM with no sharding configured behaves identically to `--shard 0 --total-shards 1`. In fact having the metric even for implicit (no-)sharding configuration is one of the benefits.

Is this a reasonable change https://github.com/kubernetes/kube-state-metrics/pull/1344/files#diff-8de36acb4313849c6ece61305374575fa2f18bdb6113d44022ee9749785d4af2R80 to resolve this?

Reply:
Registering the metrics is fine I think, calling `.WithSharding(0, 1)` does seem a bit odd here, just in terms of placement of this call.

Reply:
Ok, I've removed it again. Honestly I can't think of any way that would decouple sharding and metrics without overly complicating the code. The downside of current implementation (without the explicit `WithSharding(0,1)` call) is that if for whatever reason the `WithSharding` is not called later on, `kube_state_metrics_total_shards` will be set to `0` and incorrect. But hopefully this is rather academic; the builder is only used once in this specific file for this component.

Reply:
Yes there doesn't seem to be a smooth solution agreed. I am fine with this one, I believe its better than the alternative.